server/api.go

```diff
@@ -305,6 +305,7 @@ func (a *apiServer) Publish(ctx context.Context, req *client.PublishRequest) (
 			AckInbox:      req.AckInbox,
 			CorrelationId: req.CorrelationId,
 			AckPolicy:     req.AckPolicy,
+			Offset:        req.ExpectedOffset,
```

Comment:
After reading the code again I'd go for renaming `ensureStreamNotReadonly` to `ensurePublishPreconditions` and passing the whole request to that (can't add a comment to that part of the code, it's called in line 285 above). In there you can check whether concurrency control is enabled on the stream and require an `AckPolicy` to be set. There's also already a async error for these kinds of caller errors: `PublishAsyncError_BAD_REQUEST`.